include/mbgl/style/sources/geojson_source.hpp

```diff
@@ -1,17 +1,46 @@
 #pragma once
 
 #include <mbgl/style/source.hpp>
-#include <mbgl/util/geojson.hpp>
+#include <mapbox/geojson.hpp>
+
+namespace mapbox {
+
+namespace geojsonvt {
+class GeoJSONVT;
+} // namespace geojsonvt
+
+namespace supercluster {
+class Supercluster;
+} // namespace supercluster
+
+} // namespace mapbox
 
 namespace mbgl {
 namespace style {
 
+using GeoJSONVTPointer = std::unique_ptr<mapbox::geojsonvt::GeoJSONVT>;
+using SuperclusterPointer = std::unique_ptr<mapbox::supercluster::Supercluster>;
+
+struct GeoJSONOptions {
+    // GeoJSON-VT options
+    uint8_t maxzoom = 18;
+    uint16_t buffer = 128;
+    double tolerance = 0.375;
+
+    // Supercluster options
+    bool cluster = false;
+    uint16_t clusterRadius = 50;
+    uint8_t clusterMaxZoom = 17;
+};
+
 class GeoJSONSource : public Source {
 public:
-    GeoJSONSource(const std::string& id);
+    GeoJSONSource(const std::string& id, const GeoJSONOptions& options_ = GeoJSONOptions());
 
     void setURL(const std::string& url);
-    void setGeoJSON(GeoJSON&&);
+    void setGeoJSON(const mapbox::geojson::geojson&);
+
+    std::string getURL();
```

Comment:
Since a GeoJSON source might not have a URL, let's make this `optional<std::string> getURL()` so that downstream code has to explicitly handle the non-present case. Conveys the intent better than a possibly-empty string.